### PR TITLE
1708 - Workplace Clean - convert set_column_bounds() and cast_to_int() to Polars

### PR DESCRIPTION
**Pull request by garypolzin**
## Description
Trello ticket [#1708](https://trello.com/c/t0NzxVrT/1708-clean-workplace10-write-polars-code-for-setcolumnbounds-and-call-in-job)

This PR is to convert the PySpark cast_to_int() and set_column_bounds() functions to Polars.

**cast_to_int()**
Previously this was a function but is now inline and takes a list of integers. `strict=False` is set to return null values for invalid strings, such as empty " "

**set_column_bounds()**
This was a fairly complex function in PySpark with the option to have upper and/or lower bound. It's only used in the ASCWDS workplace job and only uses a minimum of 1 so I've just written this inline as it's a simple when, then, otherwise clause

## Testing
- [X] Unit tests passing
- [X] Successful [Step Function run](https://eu-west-2.console.aws.amazon.com/states/home?region=eu-west-2#/v2/executions/details/arn:aws:states:eu-west-2:856699698263:execution:1708-set-col-bounds-Transform-ASCWDS-Data:c0f87c92-f600-45fb-a189-1b01a246b62e)
- [X] Outputs checked in Athena
- [X] Output Screenshots added to Trello ticket

## Migration to polars checklist
- [X] Check that similar utils functions don't already exist, or if one can be created
- [X] Check that utils are being saved in the appropriate place
- [X] Used LazyFrame option for test data
- [X] Unit tests added
- [X] Docstrings added
- [X] Added comment to pyspark functions which have been migrated with the link to the new location
      `# converted to polars -> filepath.py`
- [X] Updated CHANGELOG
- [X] Code reviewed by AI
- [X] Moved Trello ticket to PR column

## Reviewer checklist
- [ ] PR solves the Trello ticket requirements
- [ ] Outputs appear reasonable and understandable
- [ ] The overall approach is appropriate and not over-engineered
- [ ] No obvious scalability, performance or interdependency concerns
- [ ] Tests appropriately cover the main behaviour/edge cases
- [ ] Naming and structure are broadly understandable and consistent
- [ ] Docstrings are sufficient for future maintenance
- [ ] Docstrings cover non-obvious behaviour

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -19,7 +19,7 @@ parameters:
     default: ""
   terraform_version:
     type: string
-    default: 1.13.5
+    default: 1.15.7
```
Comment (RoyPriceCS50) [inline .circleci/config.yml]: Separate PR?

File: projects/_01_ingest/ascwds/fargate/clean_ascwds_workplace.py [hunk 2]
```diff
@@ -95,24 +97,18 @@ def main(
...
-    # trello 1708
-    # ascwds_workplace_df = cUtils.set_column_bounds(
-    #     ascwds_workplace_df,
-    #     AWPClean.worker_records,
-    #     AWPClean.worker_records_bounded,
-    #     AscwdsScaleVariableLimits.worker_records_lower_limit,
-    # )
+    lf = lf.with_columns(pl.col(INT_COLUMNS).cast(pl.Int32, strict=False))
+
+    lf = lf.with_columns(
```
Comment (RoyPriceCS50) [inline projects/_01_ingest/ascwds/fargate/clean_ascwds_workplace.py:103]: set_column_bounds is called inside of set_bounds_for_columns.

That loops through bounding columns in cleaning capacity tracker data.

File: .circleci/config.yml (hunk 1 above)
Comment (garypolzin) [reply .circleci/config.yml]: ah yea that was just temporary to see if it helped - it didn't
I've reverted it back

File: projects/_01_ingest/ascwds/fargate/clean_ascwds_workplace.py (hunk 2 above)
Comment (RoyPriceCS50) [reply projects/_01_ingest/ascwds/fargate/clean_ascwds_workplace.py:103]: I think you can use the same column expansion idea as line 100.

Something like:

lf = lf.withColumns(
    pl.when(pl.col([LIST_OF_COLUMNS]) >= MIN_VALID_WORKFORCE_COUNT)
        .then(pl.col([LIST_OF_COLUMNS]))
        .otherwise(None)
        .name.suffix("_bounded")

Comment (garypolzin) [reply projects/_01_ingest/ascwds/fargate/clean_ascwds_workplace.py:103]: ah yea I spotted that, it's a strange combo of functions which you'd completely rewrite in polars. The CT version has lower and upper bounds whereas this is just lower - essentially you'd be overcomplicating a function to cater for several optional scenarios when it's simple/readable enough to just put it inline as is

Comment (RoyPriceCS50) [reply projects/_01_ingest/ascwds/fargate/clean_ascwds_workplace.py:103]: It can be done like this:

class TestColumnBoundingLoop:
    def test_idea(self):
        test_lf = pl.LazyFrame(
            {
                AWPClean.total_staff: [0, 1, 5],
                AWPClean.worker_records: [0, 1, 5],
            }
        )

        returned_lf = test_lf.with_columns(
            pl.when(pl.col([AWPClean.total_staff, AWPClean.worker_records]) >= 1)
            .then(pl.col([AWPClean.total_staff, AWPClean.worker_records]))
            .otherwise(None)
            .name.suffix("_bounded")
        )

        expected_lf = pl.LazyFrame(
            {
                AWPClean.total_staff: [0, 1, 5],
                AWPClean.worker_records: [0, 1, 5],
                "totalstaff_bounded": [None, 1, 5],
                "wkrrecs_bounded": [None, 1, 5],
            }
        )

        pl_testing.assert_frame_equal(returned_lf, expected_lf)

Comment (RoyPriceCS50) [reply projects/_01_ingest/ascwds/fargate/clean_ascwds_workplace.py:103]: ooof sorry about formatting

Comment (garypolzin) [reply projects/_01_ingest/ascwds/fargate/clean_ascwds_workplace.py:103]: ah that's nice, didn't know you could do that!

**Review by RoyPriceCS50 — APPROVED** [review]
Looks good.
